File: 2022-2023/sourcecode/lab/lab5/MyFinanceAppJava/app/src/main/java/pl/edu/uwr/pum/myfinanceappjava/util/UiSetup.java
package pl.edu.uwr.pum.myfinanceappjava.util;

import android.content.Context;
import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;

import androidx.core.content.ContextCompat;
import androidx.interpolator.view.animation.FastOutSlowInInterpolator;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import java.util.Objects;

import pl.edu.uwr.pum.myfinanceappjava.R;

public final class UiSetup {

    private UiSetup(){}

    private static final float tabScaleLow = 0.7f;
    private static final float tabScaleHigh = 1f;

    private static final int[] tabTitles = {R.string.overview, R.string.accounts, R.string.bills};
    private static final int[] tabIcons = {R.drawable.ic_overview, R.drawable.ic_accounts, R.drawable.ic_bills};

    public static void tabLayoutSetup(Context context, TabLayout tabLayout, ViewPager2 viewPager) {
        setupTabMediator(context, tabLayout, viewPager);
        ViewGroup vg = (ViewGroup) tabLayout.getChildAt(0);
        initialTabsSetup(tabLayout, vg);
        setupTabSelection(context, tabLayout, vg);
    }

    private static void setupTabSelection(Context context, TabLayout tabLayout, ViewGroup vg) {
        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                tab.setText(context.getString(tabTitles[tab.getPosition()]));
                Objects.requireNonNull(tab.getIcon()).setTint(Color.WHITE);
                
                setupAnimation(vg.getChildAt(tab.getPosition()), tabScaleHigh,
                        (long) context.getResources().getInteger(android.R.integer.config_mediumAnimTime));
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                tab.setText("");
                setupAnimation(vg.getChildAt(tab.getPosition()), tabScaleLow, 1L);
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });
    }

    private static void setupAnimation(View view, float scale, long duration) {
        view.animate()
                .scaleX(scale)
                .scaleY(scale)
                .setInterpolator(new FastOutSlowInInterpolator())
                .setDuration(duration)
                .start();
    }

    private static void initialTabsSetup(TabLayout tabLayout, ViewGroup vg) {
        for (int i = 0; i < vg.getChildCount(); i++) {
            if (i == tabLayout.getSelectedTabPosition())
                continue;
            View a = vg.getChildAt(i);
            a.setScaleX(tabScaleLow);
            a.setScaleY(tabScaleLow);
        }
    }

    private static void setupTabMediator(Context context, TabLayout tabLayout, ViewPager2 viewPager ){
        new TabLayoutMediator(tabLayout, viewPager,
                (tab, position) -> {
                    tab.setIcon(ContextCompat.getDrawable(context, tabIcons[position]));
                    if (position == 0) {
                        tab.setText(context.getString(tabTitles[position]));
                        Objects.requireNonNull(tab.getIcon()).setTint(Color.WHITE);
                    }
                }
        ).attach();
    }
}

File: 2022-2023/sourcecode/lab/lab5/MyFinanceAppKotlin/app/src/main/java/pl/edu/uwr/pum/myfinanceappkotlin/util/UiSetup.kt
package pl.edu.uwr.pum.myfinanceappkotlin.util

import android.content.Context
import android.graphics.Color
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import pl.edu.uwr.pum.myfinanceappkotlin.R

private const val tabScaleLow = 0.7f
private const val tabScaleHigh = 1f

private val tabTitles = listOf(R.string.overview_name, R.string.accounts_name, R.string.bills_name)
private val tabIcons = listOf(R.drawable.ic_overview, R.drawable.ic_accounts, R.drawable.ic_bills)

fun tabLayoutSetup(context: Context, tabLayout: TabLayout, viewPager2: ViewPager2) {
    val vg = tabLayout.getChildAt(0) as ViewGroup
    TabLayoutMediator(tabLayout, viewPager2) { tab, position ->
        tab.icon = ContextCompat.getDrawable(context, tabIcons[position])
        if (position == 0) tab.apply {
            text = context.getString(tabTitles[position])
            icon?.setTint(Color.WHITE)
        }
    }.attach()

    for (i in 0 until vg.childCount) {
        if (i == tabLayout.selectedTabPosition) continue
        val a = vg.getChildAt(i)
        a.scaleX = tabScaleLow
        a.scaleY = tabScaleLow
    }

    tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
        override fun onTabSelected(tab: TabLayout.Tab) {
            tab.apply {
                text = context.getString(tabTitles[tab.position])
                icon?.setTint(Color.WHITE)
            }

            val a = vg.getChildAt(tab.position)
            a.animate()
                .scaleX(tabScaleHigh)
                .scaleY(tabScaleHigh)
                .setInterpolator(FastOutSlowInInterpolator())
                .setDuration(
                    context.resources.getInteger(android.R.integer.config_mediumAnimTime).toLong()
                )
                .start()
        }

        override fun onTabUnselected(tab: TabLayout.Tab) {
            tab.text = ""
            val a = vg.getChildAt(tab.position)
            a.animate()
                .scaleX(tabScaleLow)
                .scaleY(tabScaleLow)
                .setInterpolator(FastOutSlowInInterpolator())
                .setDuration(1L)
                .start()
        }

        override fun onTabReselected(tab: TabLayout.Tab) {}
    })
}